core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import cats.implicits._
+import rankn._
+
+import Identifier.Constructor
+
+sealed abstract class NormalExpression {
+  /*
+   * maxLambdaVar is to keep track of what is the largest de bruijn index
+   * of lambda variables in the expression. This is useful because if this number
+   * is positive then there are unbound variables and it should not be cached (unless
+   * you want to be clever about cacheing values for when they are bound in an outer
+   * scope). And when they are negative it implies there are eta reduction opportunities.
+   * None essentially means -Infinity as either there are no linked expressions
+   * or there are no lambda variables used in the linked expressions
+   */
+  def maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(NormalPattern, NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  /*
+   * It is reasonable to ask how you can define a lambda without an identifier
+   * for its argument in its expression. This is a benefit of de bruijn indexing.
+   * When all lambdas have exactly one argument you identify the var by how many
+   * lambdas out you have to travel.
+   *
+   * eg \x -> \y -> [y,x] would have normalization Lambda(Lambda(Apply(Apply(List, LambdaVar(1)), LambdaVar(0))))
+   *
+   * ref: https://en.wikipedia.org/wiki/De_Bruijn_index
+   */
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
```

Comment:
oh, I thought the `def` was only important on the abstract class. I am totally prematurely optimizing here though so I'll just make it a `def` instead of over thinking it